### PR TITLE
Parallelize snapshot creation

### LINKED ISSUE #58: Parallelize snapshot creation
# 🚀 Feature Proposal

For the [weekly reporting process](https://github.com/cisagov/cyhy-reports/blob/develop/extras/create_snapshots_reports_scorecard.py), create snapshots in parallel, rather than serially.

## Motivation

As of today, creating snapshots is the biggest chunk of our weekly reporting time (about 700 minutes out of 850 minutes total; that's over 80% of the total time and climbing with each new report recipient we add).  By creating multiple snapshots at once, we should be able to significantly reduce the total time it takes to create snapshots, just like we have done with report creation.

Since we have to pause the commander while snapshots and reports are generated, shorter reporting time means more weekly uptime for CyHy processing.

### PR DESCRIPTION
**Pull request by dav3r**
#  #

## 🗣 Description ##
This PR modifies the [weekly reporting script](https://github.com/cisagov/cyhy-reports/blob/develop/extras/create_snapshots_reports_scorecard.py) so that snapshot creation is multithreaded, instead of single-threaded.

While I was making changes to this script, I took the opportunity to clean up some function names, variable names, comments, and logging statements to make them clearer and more consistent (https://github.com/cisagov/cyhy-reports/commit/69795cc48585b7bca5b9b6a0d0dbc6a12b23da6b, https://github.com/cisagov/cyhy-reports/commit/fd426e517a6cc034ccd097990136511772960d48, https://github.com/cisagov/cyhy-reports/commit/d208d94dfa22688a48ac4885654e01716edc704f).  I also removed some old testing code that had been commented-out, but is no longer needed (https://github.com/cisagov/cyhy-reports/commit/70ea90c3b85c19005190a3861ef29a2470589775).

Resolves #58.

## 💭 Motivation and Context ##
From #58:
> Creating snapshots is the biggest chunk of our weekly reporting time (about 700 minutes out of 850 minutes total; that's over 80% of the total time and climbing with each new report recipient we add). By creating multiple snapshots at once, we should be able to significantly reduce the total time it takes to create snapshots, just like we have done with report creation.
> 
> Since we have to pause the commander while snapshots and reports are generated, shorter reporting time means more weekly uptime for CyHy processing.

This PR starts us off with 16 threads for snapshot creation, since that is how many we currently use for report creation.  That number can be increased or decreased in Production after we observe how long the threaded snapshot creation process takes and if there are any database performance issues.  

Related, we may be able to increase the number of report-creation threads and reduce the time it takes to make the reports (currently 133 minutes), since the initial number of 16 threads predates our AWS infrastructure (we currently use an instance with 8 cores to generate reports).

Note that third-party snapshot creation (which must occur after "regular" snapshot creation) has not moved to threads yet.  Those third-party snapshots are created relatively quickly (less than 6 minutes currently), so we won't save all that much time by creating them in parallel.  Still, I added [a TODO to do that eventually](https://github.com/cisagov/cyhy-reports/commit/ec9525f26d8114edf2a05da8edaa00e9ba1645f4).

## 🧪 Testing ##

I tested these changes on my local system using the set of Production organizations, except I neutered the code so that no actual snapshots or reports were generated.  

Once I was satisfied with that testing, I created a set of 37 dummy orgs (32 "regular" orgs and 5 "third-party reporting" orgs) and imported them to a test environment (thanks @mcdonnnj for letting me use your environment).  I then ran the entire script (well, I commented out a couple of unrelated bits of code that would've required more test environment setup than I was willing to do) and confirmed that the new multithreaded code successfully generated snapshots for the regular orgs and the third-party snapshots were also generated successfully (non-threaded, but using code that was modified in this PR).

## ✅ Checklist ##

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] _All_ future TODOs are captured in issues, which are referenced in code comments.
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All new and existing tests pass.

### CODE REVIEW

File: extras/create_snapshots_reports_scorecard.py [hunk 1]
```diff
@@ -215,49 +225,147 @@ def create_weekly_snapshots(db, cyhy_db_section):
...
+    # Create the list of snapshots to generate by removing
+    # report_org_descendants (their snapshots will be created when their
+    # parent org's snapshot is created)
+    return sorted(list(set(reports_to_generate) - report_org_descendants))
+
+
+def make_list_chunks(my_list, num_chunks):
+    """Split a list into a specified number of smaller lists."""
+    for i in range(0, num_chunks):
+        yield my_list[i::num_chunks]
```
Comment (jsf9k) [inline extras/create_snapshots_reports_scorecard.py:253]: :+1:

File: extras/create_snapshots_reports_scorecard.py [hunk 2]
```diff
@@ -215,49 +225,147 @@ def create_weekly_snapshots(db, cyhy_db_section):
...
+    # Lists of orgs for each snapshot thread to process
+    snapshots_to_generate = list(
+        make_list_chunks(snapshots_to_generate, SNAPSHOT_THREADS)
+    )
+
+    # Start up the threads to create snapshots
+    for orgs in snapshots_to_generate:
+        try:
+            snapshot_thread = threading.Thread(
+                target=create_snapshots, args=(orgs, db, cyhy_db_section),
```
Comment (jsf9k) [inline extras/create_snapshots_reports_scorecard.py]: I don't think it needs to be changed right now, but this is not the most efficient way to parallelize this process.  Once a thread completes its assigned work it will sit idle.  Better would be to assign threads a single org at a time and give them new work to do (another org) when they are finished.  This way no threads would sit idle until there is no more work to be doled out.

Given that all threads are on the same instance in this case, there probably wouldn't be a lot of savings by rewriting this unless some threads are assigned groups of orgs whose snapshot creation takes considerably less time than others.  If we ever move to a Lambda or Kubernetes type of architecture for this processing, where there is heterogeneous hardware, then we would definitely want to rewrite the parallelization to handle a single org at a time.

**Review by jsf9k — APPROVED** [review]
This looks good to me.  I did leave a comment describing an improvement; we may want to create an issue from that.

File: extras/create_snapshots_reports_scorecard.py (hunk 2 above)
Comment (hillaryj) [reply extras/create_snapshots_reports_scorecard.py]: I added this info to the [cyhy-2.0 requirements issue](https://github.com/cisagov/cyhy-system/issues/7#issuecomment-759478386); this doesn't obviate creating an issue in this repo for this as a future improvement.

**Review by hillaryj — APPROVED** [review]
🎉

File: extras/create_snapshots_reports_scorecard.py (hunk 2 above)
Comment (jsf9k) [reply extras/create_snapshots_reports_scorecard.py]: I also created #62 for the parallelization improvement work.  I think it's good to mention it in both places.

File: extras/create_snapshots_reports_scorecard.py [hunk 3]
```diff
@@ -215,49 +225,147 @@ def create_weekly_snapshots(db, cyhy_db_section):
...
             )
-        else:
-            failure_snaps.append(i)
-            logging.info("Added to failed snaps: %s", i)
-            logging.info("Stderr failure detail: %s%s", data, err)
+            snapshot_threads.append(snapshot_thread)
+            snapshot_thread.start()
+        except:
```
Comment (mcdonnnj) [inline extras/create_snapshots_reports_scorecard.py]: We shouldn't catch `BaseException`.

```suggestion
        except Exception:
```

File: extras/create_snapshots_reports_scorecard.py [hunk 4]
```diff
@@ -707,16 +786,12 @@ def main():
...
     cyhy_db_section = args["CYHY_DB_SECTION"]
     scan_db_section = args["SCAN_DB_SECTION"]
     use_docker = 1
-    success_snaps = list()
-    failed_snaps = list()
-    reports_generated = list()
-    reports_failed = list()
     # To track third-party snapshot and report status
     successful_tp_snaps = list()
     failed_tp_snaps = list()
```
Comment (mcdonnnj) [inline extras/create_snapshots_reports_scorecard.py:797]: Checked against the LGTM alert, and these are definitely only used after being reassigned in:

https://github.com/cisagov/cyhy-reports/blob/5bed8506c25428b66ffe6290fdac89b8a393e47b/extras/create_snapshots_reports_scorecard.py#L924-L943

We also never do anything with the value(s) in `failed_tp_snaps` that are returned from `generate_third_party_reports()`, but that is outside the scope of this PR.

```suggestion
```

File: extras/create_snapshots_reports_scorecard.py (hunk 1 above)
Comment (mcdonnnj) [inline extras/create_snapshots_reports_scorecard.py:253]: 👍

File: extras/create_snapshots_reports_scorecard.py [hunk 5]
```diff
@@ -215,49 +225,147 @@ def create_weekly_snapshots(db, cyhy_db_section):
...
+def create_snapshot(db, cyhy_db_section, org_id, use_only_existing_snapshots):
+    """Create a snapshot for a specified organization."""
+    snapshot_start_time = time.time()
+
+    snapshot_command = ["cyhy-snapshot", "--section", cyhy_db_section, "create"]
+
+    if use_only_existing_snapshots:
+        snapshot_command.extend(["--use-only-existing-snapshots", org_id])
+    else:
+        snapshot_command.append(org_id)
```
Comment (mcdonnnj) [inline extras/create_snapshots_reports_scorecard.py]: Just a thought to reduce tracking a reference twice.

```suggestion
    if use_only_existing_snapshots:
        snapshot_command.append("--use-only-existing-snapshots")

    snapshot_command.append(org_id)
```

File: extras/create_snapshots_reports_scorecard.py [hunk 6]
```diff
@@ -215,49 +225,147 @@ def create_weekly_snapshots(db, cyhy_db_section):
...
+            if org_descendants and not use_only_existing_snapshots:
+                logging.error(
+                    " - Unsuccessful descendant snapshot(s): %s", org_descendants,
+                )
+                failed_snapshots.extend(org_descendants)
+        logging.error("Stderr failure detail: %s %s", data, err)
+    return snapshot_process.returncode
+
+
+def create_snapshots(org_list, db, cyhy_db_section):
```
Comment (mcdonnnj) [inline extras/create_snapshots_reports_scorecard.py]: I'm not a fan of `create_snapshot()`/`create_snapshots()` because of their similarity. Would something like `create_snapshots_from_list()` be more at-a-glance readable?

File: extras/create_snapshots_reports_scorecard.py [hunk 7]
```diff
@@ -215,49 +225,147 @@ def create_weekly_snapshots(db, cyhy_db_section):
...
+                target=create_snapshots, args=(orgs, db, cyhy_db_section),
             )
-        else:
-            failure_snaps.append(i)
-            logging.info("Added to failed snaps: %s", i)
-            logging.info("Stderr failure detail: %s%s", data, err)
+            snapshot_threads.append(snapshot_thread)
+            snapshot_thread.start()
+        except:
+            print("Error: Unable to start snapshot thread for %s", orgs)
```
Comment (mcdonnnj) [inline extras/create_snapshots_reports_scorecard.py]: Is there a reason we don't want to use `logging.error()` for this?

File: extras/create_snapshots_reports_scorecard.py [hunk 8]
```diff
@@ -215,49 +225,147 @@ def create_weekly_snapshots(db, cyhy_db_section):
...
+
+    snapshot_durations.sort(key=lambda tup: tup[1], reverse=True)
     logging.info("Longest Snapshots:")
-    for i in longest_snaps[:10]:
-        logging.info("%s: %s seconds", i[0], str(round(i[1], 1)))
+    for i in snapshot_durations[:10]:
+        logging.info("%s: %.1f seconds", i[0], i[1])
     logging.info(
-        "Time to complete snapshots: %.2f minutes", (round(time.time() - start, 1) / 60)
+        "Time to complete snapshots: %.2f minutes", (time.time() - start_time) / 60
```
Comment (mcdonnnj) [inline extras/create_snapshots_reports_scorecard.py]: Just as a preference thing, but I think end times should be when work is done and not at the tail of outputting information about the work done. Thoughts?

**Review by mcdonnnj — APPROVED** [review]
Thanks for getting this task done! 💪💪💪Hopefully it will result in greatly reduced snapshot generation runtime. I had some thoughts/suggestions, but nothing is particularly showstopping.

File: extras/create_snapshots_reports_scorecard.py (hunk 8 above)
Comment (hillaryj) [reply extras/create_snapshots_reports_scorecard.py]: I could go either way on this one:

- Since the output is labeled "Time to complete _snapshots_" then it's more accurate to have the stop time be at the end of processing instead of processing + printing
- On the other hand, the process of printing out the prior information should be fairly negligible and this saves an extra variable by being able to use `time.time()` directly (again, the impact of which is negligible on our modern hardware)

File: extras/create_snapshots_reports_scorecard.py (hunk 6 above)
Comment (hillaryj) [reply extras/create_snapshots_reports_scorecard.py]: I agree - this would increase usability to disambiguate the two and keep users from having to rely on the help text to tell them apart.

File: extras/create_snapshots_reports_scorecard.py (hunk 2 above)
Comment (dav3r) [reply extras/create_snapshots_reports_scorecard.py]: @jsf9k - I agree with you 100% here.  Thanks for creating #62.

File: extras/create_snapshots_reports_scorecard.py (hunk 4 above)
Comment (dav3r) [reply extras/create_snapshots_reports_scorecard.py:797]: I did some logging cleanup in https://github.com/cisagov/cyhy-reports/pull/61/commits/288be695151198711d8a87dc074114824462bf0c so that all of these variables are now used.  

The `successful_tp_snaps` and `failed_tp_snaps` variable initializations are required since they are now used in a `finally` block, meaning it's possible that we will never execute `create_third_party_snapshots()` (and assign those variables).

File: extras/create_snapshots_reports_scorecard.py (hunk 5 above)
Comment (dav3r) [reply extras/create_snapshots_reports_scorecard.py]: Good catch @mcdonnnj - thanks! 👍   
See https://github.com/cisagov/cyhy-reports/pull/61/commits/4024191d65651384dc1038bb4f97f41b775c62f4.

File: extras/create_snapshots_reports_scorecard.py (hunk 6 above)
Comment (dav3r) [reply extras/create_snapshots_reports_scorecard.py]: Excellent suggestion- see https://github.com/cisagov/cyhy-reports/pull/61/commits/36a4128ecbbb6f9373bc56924b964c84a316a91d.

File: extras/create_snapshots_reports_scorecard.py (hunk 7 above)
Comment (dav3r) [reply extras/create_snapshots_reports_scorecard.py]: It hadn't occurred to me when I updated this code, but you are correct (I had copied/pasted/modified the old reports code which didn't use logging throughout).  Remedied in https://github.com/cisagov/cyhy-reports/pull/61/commits/334497458e777ee1bccc9c15444717ec39cfb239.  

I will need to fix this in the reports code as well, but we can take care of that when we resolve #59.

File: extras/create_snapshots_reports_scorecard.py (hunk 8 above)
Comment (dav3r) [reply extras/create_snapshots_reports_scorecard.py]: I'm in agreement with both of you, but it's easy enough to fix this, so I went ahead and did so in https://github.com/cisagov/cyhy-reports/pull/61/commits/01e36ead3e9b0e21efbf37fc2f8583d0d366ea47.  Thanks for the feedback!

**Comment by dav3r** [comment]
I re-requested review from you all to make sure you're all good with the recent changes I made based on your excellent feedback.  Thanks!

PS: I re-ran the script in a test environment after making these changes and everything still looks good.

**Review by jsf9k — APPROVED** [review]
This all still looks great to me!  Nice improvements since the last time I reviewed!

**Review by hillaryj — APPROVED** [review]
Looking good! 🏆

**Comment by dav3r** [comment]
> This pull request **introduces 2 alerts** when merging [42d9453](https://github.com/cisagov/cyhy-reports/commit/42d94532c195877533b65e989886b00aa6dc1513) into [6b97eec](https://github.com/cisagov/cyhy-reports/commit/6b97eecad6c1e0ccead19692f63dc0350a5de605) - [view on LGTM.com](https://lgtm.com/projects/g/cisagov/cyhy-reports/rev/pr-25da86ca4a433091b41e6d818dd8d91aead0bb86)
> 
> **new alerts:**
> 
> * 2 for Variable defined multiple times

For posterity, it seems that these LGTM alerts are not taking into account the `try ... finally` block, as explained in [this comment](https://github.com/cisagov/cyhy-reports/pull/61#discussion_r561996464).

**Comment by mcdonnnj** [comment]
> 
> 
> > This pull request **introduces 2 alerts** when merging [42d9453](https://github.com/cisagov/cyhy-reports/commit/42d94532c195877533b65e989886b00aa6dc1513) into [6b97eec](https://github.com/cisagov/cyhy-reports/commit/6b97eecad6c1e0ccead19692f63dc0350a5de605) - [view on LGTM.com](https://lgtm.com/projects/g/cisagov/cyhy-reports/rev/pr-25da86ca4a433091b41e6d818dd8d91aead0bb86)
> > **new alerts:**
> > 
> > * 2 for Variable defined multiple times
> 
> For posterity, it seems that these LGTM alerts are not taking into account the `try ... finally` block, as explained in [this comment](https://github.com/cisagov/cyhy-reports/pull/61#discussion_r561996464).

They are assigned here
https://github.com/cisagov/cyhy-reports/blob/01e36ead3e9b0e21efbf37fc2f8583d0d366ea47/extras/create_snapshots_reports_scorecard.py#L796-L797
and then if they meet the appropriate logic, they are reassigned here
https://github.com/cisagov/cyhy-reports/blob/01e36ead3e9b0e21efbf37fc2f8583d0d366ea47/extras/create_snapshots_reports_scorecard.py#L931-L936

I think LGTM expects you do define them exclusively in that `if`/`else` logic instead of doing an initial declaration and then overriding as needed. I'm fine with how we're doing it, but wanted to give more explanation for posterity.

File: extras/create_snapshots_reports_scorecard.py (hunk 4 above)
Comment (mcdonnnj) [reply extras/create_snapshots_reports_scorecard.py:797]: LGTM is complaining about how we're solving the problem, but I'm a-ok with how we're doing it, so I'm fine with the alerts. Thanks for fleshing out the logging!

**Review by mcdonnnj — APPROVED** [review]
Looking even better to me now! Thanks again for parallelizing this job 💪💪💪

**Comment by jsf9k** [comment]
> I think LGTM expects you do define them exclusively in that `if`/`else` logic instead of doing an initial declaration and then overriding as needed. I'm fine with how we're doing it, but wanted to give more explanation for posterity.

I detest the leaky scope "feature" of Python.

**Review by felddy — APPROVED** [review]
👍 Nooice! 
d2f91f7aeb5d98d222fb083796c1fc58
